src/Fixer/Phpdoc/PhpdocParamOrderFixer.php

```diff
@@ -0,0 +1,174 @@
+                continue;
+            }
+
+            // Check for function signature
+            $functionSequence = $tokens->findSequence([[T_FUNCTION], [T_STRING], '('], $index);
+
+            if (null === $functionSequence) {
+                continue;
+            }
+
+            $funcParamTokens = $this->getFuncParameterTokens($tokens, $functionSequence);
+            $docBlock = $this->sortDocBlockParamAnnotations($tokens[$index], $funcParamTokens);
+
+            $tokens[$index] = new Token([T_DOC_COMMENT, $docBlock->getContent()]);
+        }
+    }
+
+    /**
+     * @param Tokens            $tokens
+     * @param array<int, Token> $functionSequence
+     *
+     * @return array
+     */
+    private function getFuncParameterTokens(Tokens $tokens, array $functionSequence)
+    {
+        $paramBlockStart = array_keys($functionSequence)[2];
+        $paramBlockEnd = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_PARENTHESIS_BRACE, $paramBlockStart);
+
+        $params = [];
+        for (
+            $i = $tokens->getNextTokenOfKind($paramBlockStart, [[T_VARIABLE]]);
+            null !== $i && $i < $paramBlockEnd;
+            $i = $tokens->getNextTokenOfKind($i, [[T_VARIABLE]])
+        ) {
+            $params[] = $tokens[$i];
+        }
+
+        return $params;
+    }
+
+    /**
+     * @param Token   $docBlockToken
+     * @param Token[] $funcParamTokens
+     *
+     * @return DocBlock
+     */
+    private function sortDocBlockParamAnnotations(Token $docBlockToken, array $funcParamTokens)
+    {
+        $doc = new DocBlock($docBlockToken->getContent());
+        $paramAnotations = $doc->getAnnotationsOfType('param');
+
+        // Create a list of valid @param annotations
+        $validParams = [];
+        foreach ($funcParamTokens as $funcParam) {
+            $index = $this->findParamAnnotationByIdentifier($paramAnotations, $funcParam->getContent());
+            if (is_int($index)) {
+                $validParams[$index] = $paramAnotations[$index]->getContent();
+            }
+        }
+
+        // Detect superfluous annotations
+        /** @var Annotation[] $invalidParams */
+        $invalidParams = array_diff_key($paramAnotations, $validParams);
+
+        // Reset indices
+        $validParams = array_values($validParams);
+        $invalidParams = array_values($invalidParams);
+
+        // Get annotation content and fix index
+        foreach ($invalidParams as $i => $params) {
+            unset($invalidParams[$i]);
+            $invalidParams[$i + count($validParams)] = $params->getContent();
+        }
+
+        // Concatenate the sorted annotations
+        $orderedAnnotations = $validParams + $invalidParams;
+
+        // Rewrite the param annotations
+        foreach ($paramAnotations as $i => $docAnnotation) {
+            $docAnnotation->remove();
+            $doc
+                ->getLine($docAnnotation->getStart())
+                ->setContent($orderedAnnotations[$i]);
```

Comment:
The fixer crashes on this line if I test it with a nested doc block.

I added this test case
```php
    public function testNestedPhpdoc()
    {
        $expected = <<<'EOT'
<?php

/**
 * @param string[] $array
 * @param callable $callback {
 *     @param string $value
 *     @param int    $key
 *     @param mixed  $userdata
 * }
 * @param mixed    $userdata
 *
 * @return bool
 */
function string_array_walk(array &$array, callable $callback, $userdata = null) {}
EOT;

        $input = <<<'EOT'
<?php

/**
 * @param callable $callback {
 *     @param string $value
 *     @param int    $key
 *     @param mixed  $userdata
 * }
 * @param string[] $array
 * @param mixed    $userdata
 *
 * @return bool
 */
function string_array_walk(array &$array, callable $callback, $userdata = null) {}
EOT;

        $this->doTest($expected, $input);
    }
```
And the result is this:
```
Undefined offset: 2
 php-cs-fixer/vendor/symfony/phpunit-bridge/DeprecationErrorHandler.php:108
 php-cs-fixer/src/Fixer/Phpdoc/PhpdocParamOrderFixer.php:150
 php-cs-fixer/src/Fixer/Phpdoc/PhpdocParamOrderFixer.php:79
 php-cs-fixer/src/AbstractFixer.php:75
 php-cs-fixer/tests/Test/AbstractFixerTestCase.php:127
 php-cs-fixer/tests/Fixer/Phpdoc/PhpdocParamOrderFixerTest.php:572
 php-cs-fixer/vendor/phpunitgoodpractices/traits/src/ExpectationViaCodeOverAnnotationTrait7.php:39
```

Reply:
I suggest you fix only "first level" annotations, the same way as can be seen [in PR 3911](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3911/files#diff-4aae1b68fe4f27312caade85cd01c158R115).

Reply:
Can you add a test case with `@param Foo|Bar $x`, `@param ?Foo $x` and `@param Foo[]|\Bar\Baz $x`?

Reply:
@dmvdbrugge I'll look into it.

@kubawerlos Do you mean a testcase like this (I'm a bit confused about the three `$x`)?

```php
<?php
class C {
    /**
     * @param Foo|Bar $y
     * @param ?Foo $z
     * @param Foo[]|\Bar\Baz $x
     */
    public function m($x, $y, $z) {}
}
```

Reply:
Yeah, some mix of really "creative" types.